### ISSUE TITLE
Maven Run Configuration cannot resolve source artifacts from workspace

### ISSUE THREAD

**Issue by eliasbalasis** [eclipse-m2e/m2e-core#1385]
On Eclipse 2023-03 with M2E 2.3.0 20230520-0934

executing a "Run Configuration" of "Maven Build" type.
with "Resolve Workspace artifacts" enabled,
any artifacts with source classifier in the workspace are not resolved.

The only way to make this work is forcing "package" and "install" phase using "<?m2e>" directive for all artifacts which defeats the purpose of the "Run Configuration".

**Comment by laeubi** [comment]
Can you check with latest snapshot? I remember there was a fix regarding classifier usage there.

**Comment by eliasbalasis** [comment]
I am already running the latest from https://download.eclipse.org/technology/m2e/snapshots/latest/
which seems to be 2.3.0 20230520-0934

**Comment by laeubi** [comment]
Sorry it seems it was not yet merged:
- https://github.com/eclipse-m2e/m2e-core/pull/1358

**Comment by eliasbalasis** [comment]
It seems this has been merged, but on Eclipse 2023-12 the problem is still present.
see #1628

**Comment by eliasbalasis** [comment]
This is still a problem in 2025-03

**Comment by eliasbalasis** [comment]
Thanks @laeubi 

I am confident I have tried this in the past but without success.
I will check again.

**Comment by eliasbalasis** [comment]
@laeubi , I tried forcing execution of Maven install plugin.

```
	<build>
		<plugins>
			<plugin>
				<?m2e execute onConfiguration,onIncremental?>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-source-plugin</artifactId>
			</plugin>
			<plugin>
				<?m2e execute onConfiguration,onIncremental?>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-jar-plugin</artifactId>
			</plugin>
			<plugin>
				<?m2e execute onConfiguration,onIncremental?>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-war-plugin</artifactId>
			</plugin>
			<plugin>
				<?m2e execute onConfiguration,onIncremental?>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-install-plugin</artifactId>
			</plugin>
```

Unfortunately, the forced plugins do not seem to be running on POM modules, only on JAR ans WAR modules, which is not helpful.

**Comment by laeubi** [comment]
> I am running GWT projects, which require presence of source Maven artifacts

Why is this needed? Is there a reference to the docs? As explained I can'T imagine how a source artifact could ever be mapped to a workspace project.

> I am trying to run some actions that for some reason cannot be executed by the relevant Eclipse plugins.

Can you explain why? It looks strange to want to use workspace artifacts but not execute something in the IDE... So if you really only need it for executing a maven commandline build why not simply use mvn install instead?

> However, forcing Maven install phase in Eclipse does not work for parent POM modules.

Install phase should be possible to execute, but it will just install the pom (not any "sources") but the pom itself should be something that is a workspace resolvable artifact.

> If you still need a reproduction project I could create one but it wouldn't be quick.

Yes I think a minimal example project would be good to understand the use-case and why something is not working.

**Comment by eliasbalasis** [comment]
I think I have confused you unintentionally.

By source module I mean JAR module with "sources" classifier.

Such modules produce two artifacts using `maven-source-plugin`, one with classes and one with sources and I was expecting that M2E would be able to differentiate and resolve in the Maven launcher, but when running a Maven launcher with "Resolve Workspace Artifacts" it seems only the binary artifact is discovered from the workspace, even when execution of `maven-source-plugin` is forced using `<?m2e execute onConfiguration,onIncremental?>`

> Could not find artifact `<groupId>:<artifactId>:jar:sources:<version>`

**Comment by eliasbalasis** [comment]
I am sensing some misunderstanding as well.

I understand that M2E resolves to the life project in the workspace (usually the `target` folder of it), which is where the binary JARs are loaded from when using "Resolve Workspace Artifacts".

I was expecting the same for the classifier, since the source JARs are also present in the `target` folder.

Is my expectation correct?

**Comment by laeubi** [comment]
> I understand that M2E resolves to the life project in the workspace (usually the target folder of it), which is where the binary JARs are loaded from when using "Resolve Workspace Artifacts".

No, it literally resolves to the `target/classes` folder not any jars in there.

So without workspace resolution maven would look into `~/.m2/repository/G/A/V/artifact.jar` with workspace resolution maven will be looking to `<my workspace project path>/target/classes` for a given GAV that is a workspace project.

Now assume it looks for GAV+C (e.g. "sources") maven will look at `~/.m2/repository/G/A/V/artifact-sources.jar` because m2e has nothing that would match a classifier for  a given GAV (this only works for the project main artifact).

**Comment by eliasbalasis** [comment]
> it literally resolves to the target/classes folder not any jars in there.

Good point.
It explains a lot.

Where would M2E look for `source` then?
I am guessing in local repository as you seem to be suggesting.

**Comment by eliasbalasis** [comment]
Thanks, I understand, it is impossible to resolve `sources` using "Resolve Workspace Artifacts".
